karabo_data/geometry2.py

```diff
@@ -0,0 +1,483 @@
+          The last two dimensions represent pixel y and x in the detector space.
+        centre : ndarray
+          (x, y) pixel location of the detector centre in this geometry.
+        """
+        assert data.shape == (16, 512, 128)
+        size_xy, centre = self._plotting_dimensions()
+        size_yx = size_xy[::-1]
+        tmp = np.empty((16 * 8,) + size_yx, dtype=data.dtype)
+
+        for i, (module, mod_data) in enumerate(zip(self.modules, data)):
+            tiles_data = np.split(mod_data, 8)
+            for j, (tile, tile_data) in enumerate(zip(module, tiles_data)):
+                # We store (x, y, z), but numpy indexing, and hence affine_transform,
+                # work like [y, x]. Rearrange the numbers:
+                fs_vec_yx = tile.fs_vec[:2][::-1]
+                ss_vec_yx = tile.ss_vec[:2][::-1]
+
+                # Offset by centre to make all coordinates positive
+                corner_pos = (tile.corner_pos[:2] + centre)
+                corner_pos_yx = corner_pos[::-1]
+
+                # Make the rotation matrix
+                rotn = np.stack((ss_vec_yx, fs_vec_yx), axis=-1)
+
+                # affine_transform takes a mapping from *output* to *input*.
+                # So we reverse the forward transformation.
+                transform = np.linalg.inv(rotn)
+                offset = np.dot(rotn, corner_pos_yx)  # this seems to work, but is it right?
+
+                affine_transform(tile_data, transform, offset=offset, cval=np.nan,
+                                 output_shape=size_yx, output=tmp[i * 8 + j])
+
+        # Silence warnings about nans - we expect gaps in the result
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=RuntimeWarning)
+            out = np.nanmax(tmp, axis=0)
+
+        return out, centre
+
+    def _plotting_dimensions(self):
+        """Calculate appropriate dimensions for plotting assembled data
+
+        Returns (size_x, size_y), (centre_x, centre_y)
+        """
+        corners = []
+        for module in self.modules:
+            for tile in module:
+                corners.append(tile.corners())
+        corners = np.concatenate(corners)[:, :2]
+
+        # Find extremes, add 20 px margin
+        min_xy = corners.min(axis=0).astype(int) - 20
+        max_xy = corners.max(axis=0).astype(int) + 20
+
+        size = max_xy - min_xy
+        centre = -min_xy
+        return tuple(size), centre
+
+    def plot_data(self, modules_data):
+        """Plot data from the detector using this geometry.
+
+        Returns a matplotlib figure.
+
+        Parameters
+        ----------
+
+        modules_data : ndarray
+          Should have exactly 3 dimensions: channelno, pixel_ss, pixel_fs
+          (lengths 16, 512, 128). ss/fs are slow-scan and fast-scan.
+        """
+        from matplotlib.cm import viridis
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        from matplotlib.figure import Figure
```

Comment:
Would it be better to use pyplot instead of the OO API. I guess people (including myself) are more familiar pyplot. So I'd vote for creating the figure objects with pyplot. Or is there a reason why this is a bad idea?

Reply:
yes I'd prefer to have an additional axis keyword argument for sub-plots and other things